library/x509_crt.c

```diff
@@ -2524,6 +2529,154 @@ static int x509_crt_verify_chain(
+#elif defined(__has_include)
+#if __has_include(<sys/socket.h>)
+#include <sys/socket.h>
+#endif
+#if __has_include(<arpa/inet.h>)
+#include <arpa/inet.h>
+#endif
+#endif
+
+/* Use whether or not AF_INET6 is defined to indicate whether or not to use
+ * the platform inet_pton() or a local implementation (below).  The local
+ * implementation may be used even in cases where the platform provides
+ * inet_pton(), e.g. when there are different includes required and/or the
+ * platform implementation requires dependencies on additional libraries.
+ * Specifically, Windows requires custom includes and additional link
+ * dependencies, and Solaris requires additional link dependencies.
+ * Also, as a coarse heuristic, use the local implementation if the compiler
+ * does not support __has_include(), or if the definition of AF_INET6 is not
+ * provided by headers included (or not) via __has_include() above.
+ * MBEDTLS_TEST_SW_INET_PTON is a bypass define to force testing of this code //no-check-names
+ * despite having a platform that has inet_pton. */
+#if !defined(AF_INET6) || defined(MBEDTLS_TEST_SW_INET_PTON) //no-check-names
+/* Definition located further below to possibly reduce compiler inlining */
+static int x509_inet_pton_ipv4(const char *src, void *dst);
+
+#define li_cton(c, n) \
+    (((n) = (c) - '0') <= 9 || (((n) = ((c)&0xdf) - 'A') <= 5 ? ((n) += 10) : 0))
+
+static int x509_inet_pton_ipv6(const char *src, void *dst)
+{
+    const unsigned char *p = (const unsigned char *) src;
+    int nonzero_groups = 0, num_digits, zero_group_start = -1;
+    uint16_t addr[8];
+    do {
+        /* note: allows excess leading 0's, e.g. 1:0002:3:... */
+        uint16_t group = num_digits = 0;
+        for (uint8_t digit; num_digits < 4 && li_cton(*p, digit);
+             group <<= 4, group |= digit, ++p, ++num_digits) {
+            ;
+        }
+        if (num_digits != 0) {
+            addr[nonzero_groups++] = MBEDTLS_IS_BIG_ENDIAN ? group :
+                                     (group << 8) | (group >> 8);
+            if (*p == '\0') {
+                break;
+            } else if (*p == '.' && (nonzero_groups != 0 ||
+                                     zero_group_start != -1) &&
+                       (nonzero_groups < 7) &&
+                       /* walk back to prior ':', then parse as IPv4-mapped */
+                       (*--p == ':' || *--p == ':' ||
+                        *--p == ':' || *--p == ':') &&
+                       x509_inet_pton_ipv4((const char *) ++p,
+                                           addr + --nonzero_groups) == 0) {
+                nonzero_groups += 2;
+                p = (const unsigned char *) "";
+                break;
+            } else if (*p != ':') {
+                return -1;
+            }
+        } else {
+            if (zero_group_start != -1 || *p != ':' ||
+                ((zero_group_start = nonzero_groups) == 0 &&
+                 *++p != ':')) {
```

Comment:
Another long condition with side effects. Where possible we should keep side effects in separate statements and have conditions be simple tests with no side effects as it makes everything easier to reason about.

Here we're doing the detection of `::` in a long string of conditions at the end of which is an error. To the casual reader of the code, this appears as "if [something complicated] return an error". It looks like a complicated error condition, when in fact it is doing some positive task and the "return an error" is a catchall for anything else.

It would be better to have a positive `if`-statement with an `else` that catches the errors, like:
```c
if (/* Compound condition that detects a zero-group */) {
    zero_group_start = nonzero_groups;
    ...
} else {
    return -1;
}
```

Reply:
I went for a different scheme, where the error is caught first to not go deeper into indentation :)